ehr/src/org/labkey/ehr/pipeline/GeneticCalculationsImportTask.java

```diff
@@ -272,6 +283,187 @@ else if (kinshipTable.getSqlDialect().isPostgreSQL())
         }
     }
 
+    private Double getMinCoefficient(String relation, @Nullable String relationship)
+    {
+        switch(relation)
+        {
+            case "dam":
+            case "sire":
+            case "Offspring":
+                return 0.25;
+            case "MaternalGranddam":
+            case "MaternalGrandsire":
+            case "PaternalGranddam":
+            case "PaternalGrandsire":
+                return 0.125;
+            case "Sibling":
+                if (relationship != null)
+                {
+                    if (relationship.equalsIgnoreCase("Full Sib"))
+                        return 0.25;
+
+                    return 0.125;
+                }
+        }
+        return 0.0;
+    }
+
+    // This function checks if missing coefficients are due to relations across species which are not tracked for kinship
+    // coefficients. Otherwise throw a pipeline error.
+    private void handleMissingCoefficient(User user, Container container, String id, String id2, String relation)
+    {
+        UserSchema studySchema = QueryService.get().getUserSchema(user, container, "study");
+        if (studySchema == null)
+        {
+            throw new IllegalStateException("Could not find schema 'study'");
+        }
+        TableInfo demographicsTable = studySchema.getTable("Demographics");
+        if (demographicsTable == null)
+        {
+            throw new IllegalStateException("Could not find query 'Demographics' in study schema");
+        }
+
+        SimpleFilter filter = new SimpleFilter(FieldKey.fromParts("Id"), List.of(id, id2), CompareType.IN);
+        TableSelector demographicsTs = new TableSelector(demographicsTable, PageFlowUtil.set("Id", "species"), filter, null);
+
+        String species = null;
+        try(Results results = demographicsTs.getResults())
+        {
+            while(results.next())
+            {
+                if (species == null)
+                {
+                    species = results.getString("species");
+                }
+                else
+                {
+                    if (!species.equals(results.getString("species")))
+                    {
+                        PipelineJob job = getJob();
+                        job.getLogger().info("Relation across species, kinship coefficent not calculated. Id: " + id + ", Id2: " + id2 + ", Relation: " + relation);
+                    }
+                    else
+                    {
+                        throw new IllegalStateException("Kinship validation failed for Id: " + id + ", Id2: " + id2 + ". Relation: " + relation + ". Kinship coefficent not found");
+                    }
+                }
+            }
+        }
+        catch (SQLException e)
+        {
+            throw new RuntimeException(e);
```

Comment:
Better to make this a RuntimeSQLException

Reply:
Updated